Add taxonomic classification for bins with gtdb-tk

skrakau opened the pull request:
I played with [GTDB-tk](https://ecogenomics.github.io/GTDBTk/index.html#) for taxonomic classification as an alternative for CAT:

- previous GTDB references stay accessible (CAT DBs only upon request currently)
- it will be clear which GTDB references will be compatible with which GTDB-tk version, i.e. it will be guaranteed that all pipelines version (with a specific GTDB-tk version) can be run at all times

Disadvantages:

- only uses marker genes, will not work for smaller bins

TODOs:

- [x] update BUSCO and make use of auto lineage functionality
- [ ] ~use `--force` to continue processing if an error occurs on a single genome (since `*.classify.tree` output would be interesting, parallelising over individual bins is likely not a good option)~
- [x] merge summary with Busco and quast summaries


(addresses https://github.com/nf-core/mag/issues/173)

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [x] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
 - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
 - [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

d4straub commented on main.nf:
```diff
@@ -571,6 +581,19 @@ workflow {
+         * GTDB-tk: taxonomic classifications using GTDB reference
+         */
+        if ( params.gtdb ){
+            // TODO apply only to medium & high quality MAGs (>=50% complete, <=10% contamination)?
```
If the GTDB database contains bacterial and archaeal references, than this would be currently not a good idea, because:
- BUSCO requires a database that is e.g. archaea or bacteria specific. When choosing the bacteria db, all values for archaea genomes will be strongly underestimated (e.g. a 100% complete archaea genome will be reported as 20-40% complete...).
- some microorganism with reduced genomes (e.g. symbiontic, pathogenic,...) might have with general BUSCOs low completeness estimates but are fine. So I would not fancy forward flaws with gene sets used to evaluate with BUSCO downstream.

skrakau replied:
-> use Busco auto lineage function

skrakau commented:
use `--force` to continue processing if an error occurs on a single genome, or parallelise over indiv. bins.

skrakau commented:
merge with Busco and quast summaries

skrakau commented:
Here another try:

- update to GTDB-Tk v1.5.0
   - contains an update regarding the handling of failed genomes
- created subworkflow for bin filtering, GTDB-Tk classification and summary generation
- added a bin filtering step based on BUSCO QC results, mandatory currently (completeness > 0.0)!
- did not use the GTDB-Tk parameter `--force` in the end
    - GTDB-Tk can handle failed genomes (without `--force`) in certain cases, e.g. if prodigal finds no gene or if the AAs fraction in the MSA is too low. These bins will be reported accordingly.
    - if it fails for other currently by GTDB-Tk unhandled reasons, e.g. because no marker gene could be detected, these bins will currently not be reported accordingly 
    - to keep control, I assume only bins with a certain completeness reported by BUSCO will be used, which should avoid this problem. If GTDB-Tk error occurs -> mag pipeline error
- changed the process `merge_quast_and_busco` to `merge_quast_busco_gtdbtk`:
   - combine BUSCO, QUAST and GTDB-Tk summaries, if >= two of those available
   - -> `results/GenomeBinning/bin_summary.tsv`
   - moved generation of `quast_summary.tsv` to separate process `QUAST_BINS_SUMMARY`
- add process `GTDBTK_SUMMARY` to create  `gtdbtk_summary.tsv`, containing all bins (also bins filtered out based on BUSCO QC and by GTDB-Tk)

A remaining question is, which results of GTDB-tk exactly to report in the summary file. Currently:
```
user_genome     classification  fastani_reference       fastani_ani     fastani_af      closest_placement_reference     closest_placement_ani   closest_placement_af    classification_method   msa_percent     red_value       warnings
```
(see https://ecogenomics.github.io/GTDBTk/files/summary.tsv.html#summary-tsv)


And I need to reduce the memory consumption and test out a few parameters for this ...

d4straub commented:
> A remaining question is, which results of GTDB-tk exactly to report in the summary file. Currently:

That is fine in my opinion. Better a little too many columns than too less information, in my opinion.

d4straub reviewed (approved):
Thats a great addition to the pipeline!

skrakau commented:
Great, thanks a lot @d4straub for feedback and reviewing!
